File: src/utils/pathfinding.ts
// ============================================================
// Dijkstra's shortest-path algorithm on the directed network
// ============================================================

import type { Edge } from '../types/network';

interface DijkstraResult {
  /** dist[nodeId] = shortest distance from source */
  dist: Map<number, number>;
  /** prev[nodeId] = the edge id used to reach nodeId on shortest path */
  prev: Map<number, number | null>;
}

/**
 * Run Dijkstra from `sourceNode` over the given directed edges.
 * Returns distance and predecessor maps.
 */
export function dijkstra(
  sourceNode: number,
  allNodeIds: number[],
  adjacency: Map<number, number[]>,
  edges: Edge[],
): DijkstraResult {
  const dist = new Map<number, number>();
  const prev = new Map<number, number | null>();
  // Simple priority queue via sorted array (fine for 24 nodes)
  const visited = new Set<number>();

  for (const id of allNodeIds) {
    dist.set(id, Infinity);
    prev.set(id, null);
  }
  dist.set(sourceNode, 0);

  const queue = [...allNodeIds];

  while (queue.length > 0) {
    // Pick unvisited node with smallest tentative distance
    queue.sort((a, b) => (dist.get(a) ?? Infinity) - (dist.get(b) ?? Infinity));
    const u = queue.shift()!;
    if (visited.has(u)) continue;
    visited.add(u);

    const outgoing = adjacency.get(u) ?? [];
    for (const edgeId of outgoing) {
      const edge = edges[edgeId];
      const v = edge.target;
      if (visited.has(v)) continue;
      const alt = (dist.get(u) ?? Infinity) + edge.weight;
      if (alt < (dist.get(v) ?? Infinity)) {
        dist.set(v, alt);
        prev.set(v, edgeId);
      }
    }
  }

  return { dist, prev };
}

/**
 * Reconstruct an ordered array of node IDs from source to target
 * using the predecessor map returned by dijkstra().
 * Returns [] if no path exists.
 */
export function reconstructPath(
  source: number,
  target: number,
  prev: Map<number, number | null>,
  edges: Edge[],
): number[] {
  const path: number[] = [];
  let current = target;

  while (current !== source) {
    const edgeId = prev.get(current);
    if (edgeId == null) return []; // no path
    const edge = edges[edgeId];
    path.unshift(current);
    current = edge.source;
  }
  path.unshift(source);
  return path;
}

/**
 * Nearest-Insertion Heuristic for multi-stop routing.
 *
 * Given a start node and a list of waypoints to visit, returns an ordered
 * sequence of those waypoints (excluding start) that approximates the
 * minimum total travel distance using shortest-path distances.
 *
 * Algorithm:
 *  1. Compute Dijkstra distances from start and each waypoint.
 *  2. Seed tour with the waypoint nearest to start.
 *  3. Repeat: find the unvisited waypoint nearest to any node already
 *     in the tour, then insert it at the position with minimum
 *     insertion cost  d(i,k) + d(k,j) − d(i,j).
 */
export function insertionHeuristicTour(
  start: number,
  waypoints: number[],
  allNodeIds: number[],
  adjacency: Map<number, number[]>,
  edges: Edge[],
): number[] {
  if (waypoints.length === 0) return [];
  if (waypoints.length === 1) return [...waypoints];

  // Pre-compute shortest-path distances from start and every waypoint
  const distFrom = new Map<number, Map<number, number>>();
  for (const src of [start, ...waypoints]) {
    distFrom.set(src, dijkstra(src, allNodeIds, adjacency, edges).dist);
  }
  const d = (a: number, b: number): number =>
    distFrom.get(a)?.get(b) ?? Infinity;

  // Seed: waypoint nearest to start
  const remaining = new Set(waypoints);
  let seedNode = waypoints[0];
  let seedDist = d(start, waypoints[0]);
  for (const w of waypoints) {
    const dist = d(start, w);
    if (dist < seedDist) { seedDist = dist; seedNode = w; }
  }
  const tour: number[] = [seedNode];
  remaining.delete(seedNode);

  // Nearest-insertion loop
  while (remaining.size > 0) {
    // Find uninserted waypoint k nearest to any node in [start, ...tour]
    const tourNodes = [start, ...tour];
    let bestK = -1;
    let bestKDist = Infinity;
    for (const k of remaining) {
      for (const t of tourNodes) {
        const dist = d(t, k);
        if (dist < bestKDist) { bestKDist = dist; bestK = k; }
      }
    }
    if (bestK === -1) break;

    // Find cheapest insertion position
    // fullSeq[i] → fullSeq[i+1] becomes fullSeq[i] → bestK → fullSeq[i+1]
    // Inserting between fullSeq[i] and fullSeq[i+1] maps to tour.splice(i, 0, bestK)
    const fullSeq = [start, ...tour];
    let bestInsertIdx = tour.length; // default: append at end
    let bestCost = d(fullSeq[fullSeq.length - 1], bestK);

    for (let i = 0; i < fullSeq.length - 1; i++) {
      const cost =
        d(fullSeq[i], bestK) +
        d(bestK, fullSeq[i + 1]) -
        d(fullSeq[i], fullSeq[i + 1]);
      if (cost < bestCost) { bestCost = cost; bestInsertIdx = i; }
    }

    tour.splice(bestInsertIdx, 0, bestK);
    remaining.delete(bestK);
  }

  return tour;
}

/**
 * Find the nearest reachable charging station from `fromNode`.
 * Returns { nodeId, path } or null if none reachable.
 */
export function nearestChargingStation(
  fromNode: number,
  chargingStationIds: number[],
  allNodeIds: number[],
  adjacency: Map<number, number[]>,
  edges: Edge[],
): { nodeId: number; path: number[] } | null {
  if (chargingStationIds.length === 0) return null;

  const { dist, prev } = dijkstra(fromNode, allNodeIds, adjacency, edges);

  let bestNode = -1;
  let bestDist = Infinity;
  for (const csId of chargingStationIds) {
    const d = dist.get(csId) ?? Infinity;
    if (d < bestDist) {
      bestDist = d;
      bestNode = csId;
    }
  }

  if (bestNode === -1 || bestDist === Infinity) return null;

  const path = reconstructPath(fromNode, bestNode, prev, edges);
  return path.length > 0 ? { nodeId: bestNode, path } : null;
}

File: backend/env/pathfinding.py
# ============================================================
# Pathfinding — Python mirror of src/utils/pathfinding.ts
#
# Tie-break rules (must match TypeScript exactly):
#   Dijkstra: equal-distance nodes → smaller node_id wins.
#   heapq pushes (dist, node_id) tuples; Python tuple comparison
#   naturally applies node_id as secondary key — matches TS behaviour
#   where queue.sort() is stable and leftmost minimum wins.
#
# Insertion heuristic tie-breaks:
#   nearest-k tie → smallest node id
#   cheapest insertion position tie → smallest (lower) index
# ============================================================

from __future__ import annotations

import heapq
import math
from typing import Optional

from backend.env.network import (
    ADJACENCY,
    ALL_NODE_IDS,
    CHARGING_STATION_IDS,
    EDGES,
    Edge,
    find_edge,
)


# ============================================================
# Dijkstra
# ============================================================

def dijkstra(
    source: int,
    all_node_ids: list[int] | None = None,
) -> tuple[dict[int, float], dict[int, Optional[int]]]:
    """
    Run Dijkstra from `source` on the Sioux Falls graph.

    Returns:
        dist: node_id → shortest distance from source (float)
        prev: node_id → edge_id used to reach it on shortest path (None = no predecessor)

    Tie-break: (dist, node_id) min-heap → smaller node_id wins on equal distance.
    This matches TS: sort()-based queue where smaller node ids appear first among
    equal-distance entries (stable sort, lower-id nodes inserted first).
    """
    if all_node_ids is None:
        all_node_ids = ALL_NODE_IDS

    dist: dict[int, float] = {n: math.inf for n in all_node_ids}
    prev: dict[int, Optional[int]] = {n: None for n in all_node_ids}
    dist[source] = 0.0

    # heap entries: (distance, node_id)
    heap: list[tuple[float, int]] = [(0.0, source)]
    visited: set[int] = set()

    while heap:
        d, u = heapq.heappop(heap)
        if u in visited:
            continue
        visited.add(u)

        for edge_id in ADJACENCY.get(u, []):
            edge = EDGES[edge_id]
            v = edge.target
            if v in visited:
                continue
            alt = d + edge.weight
            if alt < dist[v]:
                dist[v] = alt
                prev[v] = edge_id
                heapq.heappush(heap, (alt, v))

    return dist, prev


def reconstruct_path(
    source: int,
    target: int,
    prev: dict[int, Optional[int]],
) -> list[int]:
    """
    Reconstruct the node-id path from source to target using the
    predecessor map from dijkstra(). Returns [] if no path exists.
    """
    path: list[int] = []
    current = target

    while current != source:
        edge_id = prev.get(current)
        if edge_id is None:
            return []
        edge = EDGES[edge_id]
        path.append(current)
        current = edge.source

    path.append(source)
    path.reverse()
    return path


# ============================================================
# Nearest charging station
# ============================================================

def nearest_charging_station(
    from_node: int,
) -> Optional[tuple[int, list[int]]]:
    """
    Find the nearest reachable charging station from `from_node`.

    Returns (node_id, path) or None if unreachable.
    Tie-break: smaller node_id wins (Dijkstra heap secondary key).
    """
    dist, prev = dijkstra(from_node)

    best_node = -1
    best_dist = math.inf

    for cs_id in CHARGING_STATION_IDS:
        d = dist.get(cs_id, math.inf)
        if d < best_dist:
            best_dist = d
            best_node = cs_id

    if best_node == -1 or math.isinf(best_dist):
        return None

    path = reconstruct_path(from_node, best_node, prev)
    return (best_node, path) if path else None


def shortest_distance(src: int, dst: int) -> float:
    """
    Return the shortest path distance in pixels from src to dst.
    Uses Dijkstra on the Sioux Falls graph.
    Returns math.inf if dst is unreachable from src.
    Unit: px (same as edge weights).
    """
    dist_map, _ = dijkstra(src)
    return dist_map.get(dst, math.inf)


# ============================================================
# Nearest-Insertion Heuristic tour
# ============================================================

def insertion_heuristic_tour(
    start: int,
    waypoints: list[int],
) -> list[int]:
    """
    Nearest-insertion heuristic for multi-stop routing.
    Returns ordered waypoints (excluding start).

    Tie-break rules (matching TS insertionHeuristicTour):
      - Nearest k among remaining: smaller node_id wins on equal distance
      - Cheapest insertion position: lower index wins on equal cost
    """
    if not waypoints:
        return []
    if len(waypoints) == 1:
        return list(waypoints)

    # Pre-compute Dijkstra distances from start and each waypoint
    dist_from: dict[int, dict[int, float]] = {}
    for src in [start] + waypoints:
        dist_from[src], _ = dijkstra(src)

    def d(a: int, b: int) -> float:
        return dist_from.get(a, {}).get(b, math.inf)

    # Seed: waypoint nearest to start — tie-break: smallest node id
    remaining = set(waypoints)
    seed_node = min(
        waypoints,
        key=lambda w: (d(start, w), w),
    )
    tour: list[int] = [seed_node]
    remaining.remove(seed_node)

    # Nearest-insertion loop
    while remaining:
        tour_nodes = [start] + tour

        # Find uninserted waypoint k nearest to any node in tour_nodes
        # tie-break: smallest node id
        best_k = min(
            remaining,
            key=lambda k: (min(d(t, k) for t in tour_nodes), k),
        )

        # Find cheapest insertion position in [start, ...tour]
        # Insert best_k at tour index `insert_idx`
        # full_seq[insert_idx] → best_k → full_seq[insert_idx + 1]
        full_seq = [start] + tour
        n_seq = len(full_seq)

        # Default: append at end (after last element of tour)
        best_cost = d(full_seq[-1], best_k)
        best_insert_idx = len(tour)

        for i in range(n_seq - 1):
            cost = d(full_seq[i], best_k) + d(best_k, full_seq[i + 1]) - d(full_seq[i], full_seq[i + 1])
            if cost < best_cost:
                best_cost = cost
                best_insert_idx = i  # insert before tour[i] (i.e. after full_seq[i])

        tour.insert(best_insert_idx, best_k)
        remaining.remove(best_k)

    return tour


# ============================================================
# Reachability matrix (precomputed for action masking)
# ============================================================

def compute_reachability() -> list[list[bool]]:
    """
    Returns a 24×24 boolean matrix: reachable[i][j] = True iff node j is
    reachable from node i (including i itself with dist=0).
    """
    matrix: list[list[bool]] = []
    for src in ALL_NODE_IDS:
        dist, _ = dijkstra(src)
        row = [not math.isinf(dist.get(tgt, math.inf)) for tgt in ALL_NODE_IDS]
        matrix.append(row)
    return matrix


# Precomputed once at import time (24-node graph is trivial to compute)
REACHABILITY: list[list[bool]] = compute_reachability()


def max_dijkstra_dist() -> float:
    """Return the maximum finite shortest-path distance across all node pairs."""
    max_d = 0.0
    for src in ALL_NODE_IDS:
        dist, _ = dijkstra(src)
        for tgt in ALL_NODE_IDS:
            d = dist.get(tgt, math.inf)
            if not math.isinf(d) and d > max_d:
                max_d = d
    return max_d


MAX_DIST: float = max_dijkstra_dist()
